src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -131,6 +132,53 @@ public void copyFrom(MemorySegment src) {
                 base(), min(), size);
     }
 
+    @Override
+    public long mismatch(MemorySegment other) {
+        AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)other;
+        final long thisSize = this.byteSize();
+        final long thatSize = that.byteSize();
+        final long minSize = Math.min(thisSize, thatSize);
+
+        this.checkRange(0, minSize, false);
```

Comment:
This is of course correct but also a bit redundant (there's a similar issue, although to a smaller extent in `copyFrom`). Let's see how performance numbers are - basically this code can be simplified by omitting the bound check on both segments (since you already picked a size that must work on both) - so that we just need confinement + read access check.

Reply:
Yeah, I did think this too at the time, but then forgot. How about I just add a private method `checkStateAndAccess`.

Reply:
For now I think we can keep it as is, as it's cleaner and reuses existing abstractions. If this becomes a performance problem going forward we can revisit.